README.md
# hscrtmpl


## Synopsis

Haskell shell script templates (Haskell)


## Description

Templates for writing shell scripts in Haskell. Includes some useful functions
and examples of things commonly done in bash. And an example of using
optparse-applicative in a shell script.


## Getting source

- Download the cabalized source package [from Hackage](http://hackage.haskell.org/package/hscrtmpl)
- Get the source with stack: `$ stack unpack hscrtmpl`
- If you're just looking, [browse the source](https://github.com/dino-/hscrtmpl)


## Contact

Dino Morelli <[dino@ui3.info](mailto:dino@ui3.info)>


changelog.md
2.0 (2023-06-28)

  * Wrote a new args-example script
  * Added a note about specifying the Stackage resolver
  * Switched to modern cabal file instead of hpack
  * Moved Stackage resolver up to 16.11 (ghc 8.8.3)
  * Removed Setup.hs
  * Added example of custom error message with lookupEnv
  * Removed instructions for darcs
  * Removed later copyright date
  * Reorganized the functions in args-example.hs
  * Updated cabal file and README to reflect multiple scripts


1.8 (2021-09-05)

  * Moved copyright up to 2021
  * Added stack.yaml.lock to .gitignore
  * Replace forSystem function


1.7 (2019-08-22)

  * Added sequence expression example
  * Replaced a case statement with an if


1.6 (2018-10-03)

  * Moved copyright date up to 2018
  * Replaced out-of-fashion `fmap` with newfangled Applicative fmap
  * Changed to two-space indenting
  * Fixed a git source repository URL that was missed
  * Added some path split and dir examples
  * Removed unused TODO.md file
  * Moved stack resolver up to 12.11
  * Added example #! for runhaskell
  * Switched to hpack
  * Removed Vim swapfiles from .gitignore


1.5 (2016-11-18)

  * Changed license from BSD3 to ISC
  * Moved copyright year up to 2016
  * Switched build and execution from cabal/runhaskell to stack
  * Commented out example code that uses regexp, it imposes more deps
  * Some changes to avoid defaultTimeLocale confusion


1.4 (2015-07-04)

   * Added time-locale-compat for backwards compatibility


1.3 (2015-07-03)

   * Most of the commented examples are now live code
   * Replaced deprecated System.Cmd with System.Process
   * Updated copyright date
   * Fixed ambiguous import of defaultTimeLocale
   * Fixed some things in the cabal file
   * Added README.md file


1.2 (2014-02-28)

   * Updated cabal fields
   * Updated copyright date
   * Now using a shorter date format string
   * Updated date/time code for System.Time deprecation


1.1 (2013-09-13)

   * Added a Setup.hs to make this a valid cabalized package
   * Added new utility functions toExitCode, exitBool and forSystem
   * Added custom boringfile
  

1.0 (2013-04-15)

   * Initial release


args-example.hs
#! /usr/bin/env stack
-- stack runghc --package ansi-wl-pprint --package heredoc --package optparse-applicative
{-# LANGUAGE QuasiQuotes #-}

-- If you need a specific resolver, do this:
-- stack --resolver lts-16.11 runghc --package ...

{-
  This is a 'skeleton file' for writing shell scripts with Haskell that have
  sophisticated arg parsing with optparse-applicative.

  The idea here is to take a copy of this script and hack it to do what you
  need quickly. Throw the rest out.

  For more examples of doing common shell tasks, see the hscrtmpl.hs script
  also in this project.


  Dino Morelli <dino@ui3.info>
  https://github.com/dino-/hscrtmpl
  version: 2.0  2023-06-27
-}

import Options.Applicative
import System.Environment ( getProgName )
import Text.Heredoc ( here )
import Text.PrettyPrint.ANSI.Leijen ( string )
import Text.Printf ( printf )


main :: IO ()
main = do
  opts <- parseOpts
  print opts


data Verbosity = Normal | Verbose
  deriving Show  -- For debugging, possibly not needed

data Options = Options
  { optSwitch :: Bool
  , optMaybeString :: Maybe String
  , optFlag :: Verbosity
  , optString :: String
  , optArgument :: String
  , optArguments :: [String]
  }
  deriving Show  -- For debugging, possibly not needed


parseOpts :: IO Options
parseOpts = do
  pn <- getProgName
  execParser $ info (parser <**> helper)
    (  header (printf "%s - Argument parsing demo script" pn)
    <> footer'
    )


-- For more help on building parsers: https://github.com/pcapriotti/optparse-applicative
parser :: Parser Options
parser = Options
  <$> ( switch
        (  long "switch" <> short 's'
        <> help "A switch option"
        )
      )
  <*> ( optional $ strOption
        (  long "maybe-string" <> short 'm' <> metavar "STRING"
        <> help "An optional string option"
        )
      )
  <*> ( flag Normal Verbose
        (  long "verbose" <> short 'v'
        <> help "A flag option"
        )
      )
  <*> ( strOption
        (  long "string" <> short 'S' <> metavar "STRING"
        <> help "A required string option"
        )
      )
  <*> ( argument str
        (  metavar "DIR"
        <> help "A required single argument"
        )
      )
  <*> some ( argument str
        (  metavar "FILES..."
        <> help "Multiple arguments, one or more required"
        )
      )


footer' :: InfoMod a
footer' = footerDoc . Just $ string content
  where content = [here|OVERVIEW

Put more info here about what this script is for and how it works

v2.0  2023-06-27  Dino Morelli <dino@ui3.info>|]


hscrtmpl.hs
#! /usr/bin/env stack
{- stack runghc -}

-- If you need a specific resolver, do this:
-- stack --resolver lts-7.8 runghc

{- No stack? Use this #! instead of the two lines above:
#! /usr/bin/env runhaskell
-}

{-
  This is a 'skeleton file' for writing shell scripts with Haskell

  Haskell is my go-to language for anything more complicated than a simple
  sequence of shell commands in a bash script. Doing list processing and a lot
  of logic in bash is a grind, to me.

  The idea here is to take a copy of this script and hack it to do what you
  need quickly. Throw the rest out.

  The script starts off (after the imports) with a main full of examples of
  common things along with their bash counterparts.

  After that are a few functions that simplify things like getting the date as
  a String, logging a date-stamped String to stdout and manupulating an
  ExitCode as a true/false value.


  Dino Morelli <dino@ui3.info>
  https://github.com/dino-/hscrtmpl
  version: 2.0  2023-06-27
-}

import Control.Monad
import Data.List
import Data.Maybe
import Data.Time
import Data.Time.Format ( defaultTimeLocale )
-- Or use time-locale-compat for backwards compatibility with GHC < 7.10
--import Data.Time.Locale.Compat ( defaultTimeLocale )
import System.Directory
import System.Environment
import System.Exit
import System.FilePath
import System.Process
import Text.Printf
--import Text.Regex


main :: IO ()
main = do
  putStrLn "This is a shell script"

                                             -- in bash:
  -- dates (Date.Time)
  putStrLn =<< date                         -- date
  putStrLn =<< dateFormat "%Y%m%d"          -- date +"%Y%m%d"
    -- These two functions below

  -- file/dir things
  --   (System.Directory)
  putStrLn =<< getHomeDirectory             -- echo $HOME
  print =<< doesFileExist "foo"             -- [ -f foo ]
  print =<< doesDirectoryExist "bar"        -- [ -d bar ]
  putStrLn =<< getCurrentDirectory          -- pwd
  --setCurrentDirectory "/tmp"                -- cd /tmp
  print =<< makeAbsolute "some/relative/path"
  --   (System.FilePath)
  let (path, ext) = splitExtension "foo/bar.baz"
  print (path, ext)

  -- conditional statements (Control.Monad)
  e <- doesFileExist "hscrtmpl.hs"          -- [ -f hscrtmpl.hs ]
  when e $ putStrLn "This script exists!"   --    && echo "This script exists!"
   

  -- environment variables (System.Environment)

  -- This will exit with a message and failure code if the variable doesn't exist!
  putStrLn =<< getEnv "SHELL"               -- echo $SHELL

  -- To handle unset variables, use lookupEnv :: String -> IO (Maybe String)

  -- someValue <- maybe                        someValue="${SOMEVAR:?Error: SOMEVAR not set}"
  --    (die "Error: SOMEVAR is not set")
  --    pure =<< lookupEnv "SOMEVAR"

  someValue <- fromMaybe "default value"    -- someValue="${SOMEVAR:-default value}"
     <$> lookupEnv "SOMEVAR"
  putStrLn someValue


  -- arguments (System.Environment)
  --(arg1 : arg2 : _) <- getArgs              -- arg1=$1 ; arg2=$2

  -- string interpolation (Text.Printf)
  printf "The %s is %d\n"                   -- S="answer" ; D=42
    "answer" (42::Int)                     -- echo "The $S is $D"

  -- sequence expressions
  -- [1..10]                                   -- {1..10}

  -- execution, exit code (System.Process)
  ec <- system "ls -l"                      -- ls -l
  print ec                                  -- echo $?

  -- execution, capture stdout (System.Process)
  output <- readProcess "ls" ["-l"]         -- output=$(ls -l)
    "stdin data, if desired"

    -- or use readProcessWithExitCode
    -- :: FilePath -> [String] -> String -> IO (ExitCode, String, String)
    --    program     args        stdin         exitcode  stdout  stderr

  -- regular expressions (Text.Regex)
  --print $ matchRegex                        -- (see bash docs for =~
  --   (mkRegex "a(.)b(.)") "axby"            --  and BASH_REMATCH)

  -- Handy date-stamping logM function below
  logM "Example of a log message"

  -- exiting (System.Exit)
  exitSuccess                               -- exit 0
  --exitFailure                             -- exit 1
  --exitWith $ ExitFailure 3                -- exit 3


{- Get the current date/time as a string in RFC822 format
   Looks like this in my locale: Mon Feb 13 16:21:38 EST 2012
-}
date :: IO String
date = dateFormat "%c"


{- Get the current date/time as a string in the specified format
   For format string help, see man 3 strftime
-}
dateFormat :: String -> IO String
dateFormat fmt = formatTime defaultTimeLocale fmt <$>
   (getCurrentTime >>= utcToLocalZonedTime)


{- Output a message with datestamp
-}
logM :: String -> IO ()
logM msg = do
   tstamp <- dateFormat "%F %T"
   printf "%s> %s\n" tstamp msg


{- Turn an exit code (say, from system) into a Bool
-}
ok :: ExitCode -> Bool
ok ExitSuccess = True
ok _           = False


{- Turn a Bool into an exit code
-}
toExitCode :: Bool -> ExitCode
toExitCode True  = ExitSuccess
toExitCode False = ExitFailure 1


{- Exit with a success or failure code using a Bool
-}
exitBool :: Bool -> IO ()
exitBool = exitWith . toExitCode


{- This is similar to the for/in/do/done construct in bash with an important
   difference. This action evaluates to the "worst" exit code of all IO actions
   that are passed after they are evaluated.
-}
sequenceWorstEC :: [IO ExitCode] -> IO ExitCode
sequenceWorstEC as =
  fromMaybe ExitSuccess . listToMaybe . reverse . sort <$> sequence as


{- This is similar to the for/in/do/done construct in bash with an important
   difference. This action evaluates to the "worst" exit code of all shell
   commands that are passed after they are evaluated.
-}
systemWorstEC :: [String] -> IO ExitCode
systemWorstEC = sequenceWorstEC . map system
